File: backend/src/ontoforge_server/core/exceptions.py
import uuid


class OntoForgeError(Exception):
    """Base exception for OntoForge."""


class NotFoundError(OntoForgeError):
    """Raised when a requested resource is not found."""


class ConflictError(OntoForgeError):
    """Raised when an operation conflicts with existing state."""


class ValidationError(OntoForgeError):
    """Raised when request or business logic validation fails."""

    def __init__(self, message: str, details: dict | None = None):
        super().__init__(message)
        self.details = details


class CascadeRequiredError(OntoForgeError):
    """Raised when a schema change would break scoped ontologies and cascade is not enabled."""

    def __init__(self, message: str, affected_ontologies: list[str]):
        super().__init__(message)
        self.affected_ontologies = affected_ontologies


class StoreError(OntoForgeError):
    """Raised when the persistence adapter fails in a way no domain exception describes.

    The adapter raises this instead of letting a database failure escape the
    persistence port (port contract rule 4, ``core/ports.py``). It deliberately
    carries no storage detail: the originating exception is chained as
    ``__cause__`` and logged by the adapter against ``error_id``, which is the
    only thing the client receives that ties its response to that log entry.
    Nothing vendor-specific reaches the caller (decision 010).
    """

    def __init__(
        self,
        message: str = "A storage operation failed",
        error_id: str | None = None,
    ):
        super().__init__(message)
        self.error_id = error_id or uuid.uuid4().hex[:8]

File: server/src/core/exceptions.ts
/**
 * Exception taxonomy. The interface layer maps each class to its HTTP
 * status and envelope code (see `app.ts`).
 */

import { randomBytes } from "node:crypto";

/** Base class for all OntoForge domain errors. */
export class OntoForgeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = new.target.name;
  }
}

/** A requested resource does not exist. -> 404 RESOURCE_NOT_FOUND */
export class NotFoundError extends OntoForgeError {}

/** An operation conflicts with existing state. -> 409 RESOURCE_CONFLICT */
export class ConflictError extends OntoForgeError {}

/** Request or business-logic validation failed. -> 422 VALIDATION_ERROR */
export class ValidationError extends OntoForgeError {
  details: Record<string, unknown> | null;

  constructor(message: string, details: Record<string, unknown> | null = null) {
    super(message);
    this.details = details;
  }
}

/** A schema change would break scoped lenses and cascade was not
 * requested. -> 409 CASCADE_REQUIRED */
export class CascadeRequiredError extends OntoForgeError {
  affectedLenses: string[];

  constructor(message: string, affectedLenses: string[]) {
    super(message);
    this.affectedLenses = affectedLenses;
  }
}

/**
 * The persistence adapter failed in a way no domain exception describes.
 * -> 500 STORAGE_ERROR
 *
 * Deliberately carries no storage detail: the originating failure is set as
 * `cause` and logged by the adapter against `errorId`, which is the only
 * thing the client receives that ties its response to that log entry.
 * Nothing vendor-specific reaches the caller.
 */
export class StoreError extends OntoForgeError {
  errorId: string;

  constructor(message = "A storage operation failed", errorId?: string) {
    super(message);
    this.errorId = errorId ?? randomBytes(4).toString("hex");
  }
}
